tests/baselines/reference/mixinFunctionReturningThis.js

```diff
@@ -0,0 +1,86 @@
+//// [mixinFunctionReturningThis.ts]
+class A {
+  constructor(name: string) {}
+}
+
+type Constructor = new (...args: any[]) => {};
+ 
+function MixB<TBase extends Constructor>(Base: TBase) {
+  return class extends Base {
+    getThis(): this {
+      return this;
+    }
+  };
+}
+
+export default MixB(A);
+
+//// [mixinFunctionReturningThis.js]
+"use strict";
+var __extends = (this && this.__extends) || (function () {
+    var extendStatics = function (d, b) {
+        extendStatics = Object.setPrototypeOf ||
+            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
+            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
+        return extendStatics(d, b);
+    };
+    return function (d, b) {
+        if (typeof b !== "function" && b !== null)
+            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+Object.defineProperty(exports, "__esModule", { value: true });
+var A = /** @class */ (function () {
+    function A(name) {
+    }
+    return A;
+}());
+function MixB(Base) {
+    return /** @class */ (function (_super) {
+        __extends(class_1, _super);
+        function class_1() {
+            return _super !== null && _super.apply(this, arguments) || this;
+        }
+        class_1.prototype.getThis = function () {
+            return this;
+        };
+        return class_1;
+    }(Base));
+}
+exports.default = MixB(A);
+
+
+//// [mixinFunctionReturningThis.d.ts]
+declare class A {
+    constructor(name: string);
+}
+declare const _default: {
+    new (...args: any[]): {
+        getThis(): this;
+    };
+} & typeof A;
+export default _default;
+
+
+//// [DtsFileErrors]
+
+
+tests/cases/compiler/mixinFunctionReturningThis.d.ts(6,20): error TS2526: A 'this' type is available only in a non-static member of a class or interface.

```

Comment:
It doesn't seem like a good thing that this pattern appears to always produce d.ts files which fail to typecheck, yet no error is produced on the actual code itself; people will be emitting broken d.ts files without knowing anything's wrong. So, I don't think that this fix is the right one unless we decide it's okay / find a way to make this not error too.

Reply:
Would you agree that these cases shouldn't be errors? If so, I can try to look into making them not error.

Also, as a side note, it is already possible to produce 'broken' d.ts files with a slightly different pattern in 4.9.5:

https://www.typescriptlang.org/play?ts=4.9.5&ssl=24&ssc=1&pln=1&pc=1#code/MYGwhgzhAEDKAOAnAlgFwKbQN4ChrQDswBbTAXmgCJKBuPPaYAewIlUQFdhUnEAKIqQBc0NigIBzAJTYG+VAAtkEAHSDyhEujr4Avjn05UAT3iYAwizFceiaBQLoA7tD4r3YRBIgiwBYwDaALoyZAB82Lo6OABmHATcyCxwwGAg6AA8ACoAQpCY6AAeGAQAJjCWrOw2vGF8eRDoIrn5Mrj4iOioHIgEjOBQKWnIktBFJeXQDZjt+NAA+hCp6fbQAIzRc6JdsMvofEtpTYQcxABG6IhSIorKslvySqqLe6uH6ToPnd290LcQnz0DCiBhwOCK8F4qEYVmhAEFVrsjnwECgMFI6BCodBSugYmAOCBoUj0iikGh0BicEA

```
// ts
class Sprite {
  name = "";
 
  constructor(name: string) {
    this.name = name;
  }
}

type Constructor = new (...args: any[]) => {};
 
function Scale<TBase extends Constructor>(Base: TBase) {
  return class Scaling extends Base {
    _scale = 1;
 
    setScale(scale: number): this {
      this._scale = scale;
      return this;
    }
  };
}

export const A = Scale(Sprite);
export default Scale(Sprite);
```

```
// d.ts
declare class Sprite {
    name: string;
    constructor(name: string);
}
export declare const A: {
    new (...args: any[]): {
        _scale: number;
// A 'this' type is available only in a non-static member of a class or interface.ts(2526)
        setScale(scale: number): this;
    };
} & typeof Sprite;
declare const _default: {
    new (...args: any[]): {
        _scale: number;
        setScale(scale: number): any;
    };
} & typeof Sprite;
export default _default;
```

Reply:
> Would you agree that these cases shouldn't be errors? If so, I can try to look into making them not error.

I don't think I have enough context to say one way or another on this particular point; @weswigham I'm sure has a better answer. I just know that it's not correct to emit d.ts files which don't themselves check.

> Also, as a side note, it is already possible to produce 'broken' d.ts files with a slightly different pattern in 4.9.5:

If so, that is definitely not intended and should be reported as a bug if it isn't already; any case like this where we emit a d.ts file which doesn't typecheck is a bug.

Reply:
sounds good - just filed https://github.com/microsoft/TypeScript/issues/52687